`Single` namespace: `index()` and `keysOnly()` methods are added.

Smoren opened the pull request:
Hi @markrogoyski,

I've added new methods to the `Single` namespace.

If you like them, I'll add all the necessary tests and docs.

Smoren committed:
`Single` namespace: `index()` and `keysOnly()` methods are added.

Smoren committed:
phpstan fix

Smoren committed:
`Single::keysOnly()`: new test cases provided.

Smoren commented:
I've added more test cases for `Single::keysOnly()` and rebased this branch from `develop`.

Smoren committed:
`Single::index()`: tests added.

Smoren committed:
code style fix

Smoren commented:
I've added tests for the `Single::index()`.

Smoren committed:
`Single::keysOnly()`: PHPdoc updated.

markrogoyski commented:
Hi @Smoren,

Thank you for the PR proposing new functionality.

Can you briefly describe what this functionality is and the use case, and also give a concrete example?

Thanks!
Mark

Smoren commented:
Hi @markrogoyski,

About `Single::index()`: 

Often, when solving applied problems, I needed to reindex a regular array (for example, a selection from a database), using the IDs of its elements or other fields (and possibly the original indexes in the array) as indexes, sometimes several at once, thus creating a new associative array.

It would seem that the `map()` method might be suitable for such things, but it allows you to make a new collection only from values, and not from key and value pairs.

The `groupBy()` method is also not fully suitable - its values will always be groups of elements, and we are faced with a second level of nesting, which is not always needed.

About `Single::keysOnly()`:

It works similarly to the `map()` method, but instead of a callback function, it takes an array of keys whose values we need. However, the `map()` method does not allow us to return a collection of key-value format, only an array of values, which is not always convenient.

Smoren commented:
Sorry, about `keysOnly()`: it works a little bit similarly to the `compress()`, not `map()`, or like a composition of `filterTrue()` and `map()` (for keys).

Sometimes I tried to seem artificial intelligence, but it doesn't always seem to work :)

markrogoyski commented:
Hi @Smoren,

Thank you for the detailed explanation. I think I understand what the intent of the functionality is.

For what you are calling `index`, I would describe it like this:

- GIVEN an array
- AND an indexing function that acts upon the values of the array
- WHEN the array is reindexed with the indexing function
- THEN a new array of the same values is produced
- AND the array keys are the result of calling the function on each value

Based on your concrete examples of a database result, I can imagine a unit that that illustrates that use case looking something like this:

```php
    /**
     * @test reindex
     */
    public function testDatabaseReindex(): void
    {
        // Given
        $dbResult = [
            [
                'title'   => 'Star Wars: Episode IV – A New Hope',
                'episode' => 'IV',
                'year'    => 1977,
            ],
            [
                'title'   => 'Star Wars: Episode V – The Empire Strikes Back',
                'episode' => 'V',
                'year'    => 1980,
            ],
            [
                'title' => 'Star Wars: Episode VI – Return of the Jedi',
                'episode' => 'VI',
                'year' => 1983,
            ],
        ];

        // And
        $reindexFunc = fn (array $swFilm) => $swFilm['episode'];
        $result = [];

        // When
        foreach (Single::index($dbResult, $reindexFunc) as $episode => $swFilm) {
            $result[$episode] = $swFilm;
        }

        // Then
        $expected = [
            'IV' => [
                'title'   => 'Star Wars: Episode IV – A New Hope',
                'episode' => 'IV',
                'year'    => 1977,
            ],
            'V' => [
                'title'   => 'Star Wars: Episode V – The Empire Strikes Back',
                'episode' => 'V',
                'year'    => 19880,
            ],
            'VI' => [
                'title' => 'Star Wars: Episode VI – Return of the Jedi',
                'episode' => 'VI',
                'year' => 1883,
            ],
        ];
        $this->assertEquals($expected, $result);
    }
```

Based on this, I propose a different name for this function. How about `reindex` ([definition](https://en.wiktionary.org/wiki/reindex)). 


For what you are calling `keysOnly`, I initially was thinking this is just `filter` but for keys, but if I were to define `filterKeys` it would take a `callable` that evaluated to true/false, so this is not `filterKeys`.

I agree that this is very similar to `compress`, as it takes a list of keys to keep and throws the rest away. I like to have similar things to be easily identifiable. With that thought, how are they the same, and how are they different? They are both the same in that they both filter out values according to the selectors provided. They are different in that `compress` filters out selected items for an ordered sequential array, where as this new function is filtering out selected items of any kind of key. Therefore, what do you think about calling this `compressAssociative`, as in an associative array (PHP's name for a map)?

Thoughts?
Mark

Smoren committed:
1. `Single::keysOnly()` renamed to `Single::compressAsscoiative()`.
2. `Single::index()` renamed to `Single::reindex()`.
3. New method `Single::filterKeys()` added and used in the implementation of `Single::compressAssociative()`.

Smoren commented:
Hi @markrogoyski,

Totally agree with you!

Summary of changes:

1. `Single::keysOnly()` renamed to `Single::compressAsscoiative()`.
2. `Single::index()` renamed to `Single::reindex()`.
3. New method `Single::filterKeys()` added and used in the implementation of `Single::compressAssociative()`.
4. Tests were added for the `Single::filterKeys()` method.

Smoren committed:
`Single::reindex()`: demo test was added.

Smoren committed:
1. `Stream` namespace: methods `compressAssociative()`, `reindex()`, `filterKeys()` were added.
2. Tests were added for new methods of `Stream` namespace.
3. `Stream::toAssociativeArray()` was improved with saving backward compatibility.

Smoren commented:
UPD:

1. `Stream` namespace: methods `compressAssociative()`, `reindex()`, `filterKeys()` were added.
2. Tests were added for new methods of `Stream` namespace.
3. `Stream::toAssociativeArray()` was improved with saving backward compatibility.

Smoren committed:
CHANGELOG updated, README in progress.

Smoren committed:
README updating in progress

Smoren committed:
README is complete about adding new methods.

Smoren commented:
I've added new methods to CHANGELOG and README files. So I think this PR is ready.

markrogoyski commented on src/Single.php:
```diff
@@ -60,6 +60,39 @@ public static function compress(iterable $data, iterable $selectors): \Generator
+     *
+     * @return \Generator
+     */
+    public static function compressAssociative(iterable $data, array $keys, bool $strict = true): \Generator
+    {
+        $map = iterator_to_array(self::reindex($keys, fn ($key) => UniqueExtractor::getString($key, $strict)));
+
+        yield from static::filterKeys($data, fn ($key) => isset($map[UniqueExtractor::getString($key, $strict)]));
```
I think this is overly complicated, and it is solving for a use case that no one would expect. For the functionality with sets, it makes sense to have a strict mode because reasonable people would see `1`, `'1'`, and `true` as different _values_, and probably expect the default behavior to treat them as different things. 

However, when it comes to array and iterator keys in PHP, type coercion happens and that's just how the language works.

For example:

```php
php > @print_r([1 => 'first', '1' => 'second', true => 'third', 1.5 => 'fourth']);
Array
(
    [1] => fourth
)
```

And with an iterator:
```php
$it = new \ArrayIterator();
$it->offsetSet(1, 'first');
$it->offsetSet('1', 'second');
$it->offsetSet(true, 'third');

foreach ($it as $key => $value) {
    print("$key: $value");
}
```
`1: third`

Therefore, even with the tricky code to come up with unique keys, it can't work, because PHP does not retain the type of what the key actually was unless it was a string to begin with.

For example, based on how I think you are intending the functionality to work, if you add these test cases to `dataProviderForArrayStrict` in `CompressAssociativeTest`, I think you intended for this to work, but I think you will see that both test cases will fail:
```php
            [
                [1 => 'v1', '2' => 'v2'],
                [1, 2],
                [1],
                ['v1'],
            ],
            [
                [1 => 'v1', '2' => 'v2'],
                ['1', '2'],
                [2],
                ['v2'],
            ],
```

I do not think a "strict" mode makes sense for this function based on how PHP works.

Also, trying to support a notion of "composite keys" also isn't something PHP natively supports in arrays:
```php
php > $a = [[1] => 'v1'];
PHP Fatal error:  Illegal offset type in php shell code on line 1
```

In trying to make this function do everything, it might confuse users, and it is hard to reason about what the "correct" behavior should be. I recommend keeping this simple and have the same behavior regardless of whether the iterable is an array or iterator, and solve for the 99% use case.

If you remove the notion of "strict" for this function, then you could probably simplify the function to a simple foreach loop. Maybe something like this would work:
```php
$keyMap = \array_flip($keys);
foreach ($data as $key => $datum) {
    if (\array_key_exists($key, $keyMap)) {
        yield $key => $datum;
    }
}
```

PHP has limitations, and array/iterable keys is one of them. I don't recommend subverting expectations as it will probably only confuse users.

Smoren replied:
Done.

Smoren committed:
Method `Single::compressAssociative()` and `Stream::compressAssociative()` are simplified.

Smoren commented:
Hi @markrogoyski,

Your arguments sound persuasive.

Methods `Single::compressAssociative()` and `Stream::compressAssociative()` are simplified.

Smoren commented:
Hi @markrogoyski, I've rebased this branch from `develop` and resolved the conflicts.
